impeller/entity/geometry.cc

```diff
@@ -0,0 +1,159 @@
+  }
+
+  return GeometryResult{
+      .type = GetPrimitiveType(vertices_),
+      .vertex_buffer =
+          {
+              .vertex_buffer = {.buffer = buffer,
+                                .range = Range{0, total_vtx_bytes}},
+              .index_buffer = {.buffer = buffer,
+                               .range =
+                                   Range{total_vtx_bytes, total_idx_bytes}},
+              .index_count = vertices_.GetIndices().size(),
+              .index_type = IndexType::k16bit,
+          },
+      .prevent_overdraw = false,
+  };
+}
+
+GeometryResult VerticesGeometry::GetPositionColorBuffer(
+    std::shared_ptr<Allocator> device_allocator,
+    HostBuffer& host_buffer,
+    Color paint_color,
+    BlendMode blend_mode) {
+  using VS = GeometryColorPipeline::VertexShader;
+
+  if (!vertices_.IsValid()) {
+    return {};
+  }
+
+  auto vertex_count = vertices_.GetPositions().size();
+  std::vector<VS::PerVertexData> vertex_data(vertex_count);
+  {
+    const auto& positions = vertices_.GetPositions();
+    const auto& colors = vertices_.GetColors();
+    for (size_t i = 0; i < vertex_count; i++) {
+      auto color = Color::BlendColor(paint_color, colors[i], blend_mode);
+      vertex_data[i] = {
+          .position = positions[i],
+          .color = color,
+      };
+    }
+  }
+
+  size_t total_vtx_bytes = vertex_data.size() * sizeof(VS::PerVertexData);
+  size_t total_idx_bytes = vertices_.GetIndices().size() * sizeof(uint16_t);
+
+  DeviceBufferDescriptor buffer_desc;
+  buffer_desc.size = total_vtx_bytes + total_idx_bytes;
+  buffer_desc.storage_mode = StorageMode::kHostVisible;
+
+  auto buffer = device_allocator->CreateBuffer(buffer_desc);
+
+  if (!buffer->CopyHostBuffer(reinterpret_cast<uint8_t*>(vertex_data.data()),
+                              Range{0, total_vtx_bytes}, 0)) {
+    return {};
+  }
+  if (!buffer->CopyHostBuffer(reinterpret_cast<uint8_t*>(const_cast<uint16_t*>(
+                                  vertices_.GetIndices().data())),
+                              Range{0, total_idx_bytes}, total_vtx_bytes)) {
+    return {};
+  }
+
+  return GeometryResult{
+      .type = GetPrimitiveType(vertices_),
+      .vertex_buffer =
+          {
+              .vertex_buffer = {.buffer = buffer,
+                                .range = Range{0, total_vtx_bytes}},
+              .index_buffer = {.buffer = buffer,
+                               .range =
+                                   Range{total_vtx_bytes, total_idx_bytes}},
+              .index_count = vertices_.GetIndices().size(),
+              .index_type = IndexType::k16bit,
+          },
+      .prevent_overdraw = false,
+  };
+}
+
+GeometryResult VerticesGeometry::GetPositionUVBuffer(
+    std::shared_ptr<Allocator> device_allocator,
+    HostBuffer& host_buffer,
+    ISize render_target_size) {
+  // TODO(jonahwilliams): support texture coordinates in vertices.
+  return {};
+}
+
+GeometryVertexType VerticesGeometry::GetVertexType() {
```

Comment:
This is a bool in the design, but we'll need to figure out all of the vertices combinations to support first - the color sources might need to know about it.

For example, if we have texture coordinates and a tiled texture image source, blend mode, and a paint color, do we need to blend those together?

Reply:
Ah geez, I knew I was forgetting something. I need to go through and check all these cases. If we end up needing to blend in the vertex colors for images and/or gradients, I think we can easily support it by finish up adding all the blends on the GPU and basically do 23 item switch statement -- although dynamic dispatch should be possible on some backends!

Reply:
(In practice this would just look like a couple of extra uniforms and a call into a shader library function in each color source fragment shader.)